src/shims/foreign_items.rs

```diff
@@ -420,10 +420,17 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
             "miri_get_alloc_id" => {
                 let [ptr] = this.check_shim(abi, Abi::Rust, link_name, args)?;
                 let ptr = this.read_pointer(ptr)?;
-                let (alloc_id, _, _) = this.ptr_get_alloc_id(ptr)?;
+                let (prov, offset) = ptr.into_parts();
+                let alloc_id = if prov.is_some() {
+                    let (alloc_id, _, _) = this.ptr_get_alloc_id(ptr)?;
+                    alloc_id
+                } else {
+                    intptrcast::GlobalStateInner::alloc_id_from_addr(this, offset.bytes())
+                        .expect("pointer passed to miri_get_alloc_id must not be dangling")
+                };
```

Comment:
What is this supposed to achieve, compared to the previous implementation? The previous implementation looked good to me, it just lacked docs.

Reply:
I think the previous implementation was wrong because it had a debugging tool reporting UB. I suppose we could `err_machine_stop` instead?

Reply:
Should we test this failure path? Seems not strictly necessary for a debugging tool like this, but since we're already taking care thatthe message is helpful...

Reply:
I think the previous implementation was a lot better since it used the shared codepath for determining the AllocId. This is messing with internals of the pointer representation that I don't think we want to unfold here. Calling `into_parts` is a red flag IMO.

Reply:
Ah, I thought we were more opposed to messing with InterpError

Reply:
We are fairly strictly opposed to catching InterpError. But this is just replacing one error by another, and it seems strictly better than basically re-implementing `ptr_get_alloc_id`.